### PR TITLE
Documentation for new 3D diagnostic tendencies

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
The "develop" and "main" branches of UFS, FV3, and ccpp/physics now support 3D diagnostic tendencies for tracers. This PR updates the documentation to match the new method.

**Edit July 20, 2021**: The code is now in the "main" and "develop" branches, so this PR contains documentation for the latest code.

Changes written by @SamuelTrahanNOAA by request of @ligiabernardet and @climbfuji 

Code PRs are here:

https://github.com/NOAA-EMC/fv3atm/pull/332
https://github.com/NCAR/ccpp-physics/pull/681
https://github.com/ufs-community/ufs-weather-model/pull/660

### CODE REVIEW

**Comment by SamuelTrahanNOAA** [comment]
I'm uncertain of the SCM aspects of the documentation, so I'm hoping others can comment on that in their reviews.

**Comment by ligiabernardet** [comment]
It is good to review this PR, but it should only be merged when the code changes themselves go to master.

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 1]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
+
+Note that there is a fourth namelist variable, ``lssav``, associated with the output of
+parameterization-specific information. The value of ``lssav`` is overwritten to true in the code, so the
+value used in the namelist is irrelevant.
 
 While the tendencies output by the SCM are instantaneous, the tendencies output by the UFS are averaged
 over the number of hours specified by the user in variable ``fhzero`` in the ``&gfs_physics_nml`` portion of the
 namelist file ``input.nml``. Variable ``fhzero`` must be an integer (it cannot be zero). 
 
+This example namelist selects all tendencies from microphysics processes, and all tendencies of temperature.
```
Comment (ligiabernardet) [inline CCPPtechnical/source/ParamSpecificOutput.rst:91]:
```suggestion
This example namelist selects all tendencies from microphysics processes, and all tendencies of temperature. The naming convention for ``dtend_select`` is explained in the next section.
```

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 2]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
+
+Tendency Names
+--------------
+
+Tendency variables follow this naming pattern, which is used to enable calculation (``input.nml``) and output
+(``diag_table``) of the variable:
+
+.. code::
+
+   dtend_VARIABLE_PROCESS
```
Comment (ligiabernardet) [inline CCPPtechnical/source/ParamSpecificOutput.rst:111]:
```suggestion
   dtend_variable_process
```

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 3]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
+.. code:: fortran
+
+   &gfs_physics_nml
+     ldiag3d = .true. ! enable basic diagnostics
+     qdiag3d = .true. ! also enable tracer diagnostics
+     dtend_select = 'dtend_temp_nophys' ! All configurations have non-physics temperature tendencies
+     ! ... other namelist parameters ...
+   /
+
+You will see lines like this in the model's standard output stream:
```
Comment (ligiabernardet) [inline CCPPtechnical/source/ParamSpecificOutput.rst:129]:
```suggestion
After recompiling and running the model, you will see lines like this in the model's standard output stream:
```

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 4]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
+   | Oxygen Ion Concentration                        | ``o_ion``      | ``qdiag3d``    | ``dtend(:,:,dtidx(100+nto,:))``              | kg kg\ :sup:`-1` s\ :sup:`-1` |
+   +-------------------------------------------------+----------------+----------------+----------------------------------------------+-------------------------------+
+   | Oxygen Concentration                            | ``o2``         | ``qdiag3d``    | ``dtend(:,:,dtidx(100+nto2,:))``             | kg kg\ :sup:`-1` s\ :sup:`-1` |
+   +-------------------------------------------------+----------------+----------------+----------------------------------------------+-------------------------------+
+
+
+.. _avail_tend_processes:
+
+.. table:: Processes that can change non-chemical tracer and state variables. The third column is the
+           ``process`` part of ``dtend_variable_process``.
```
Comment (ligiabernardet) [inline CCPPtechnical/source/ParamSpecificOutput.rst:223]: We need some explanation as to why _temperature_ is considered a _process_, when it sounds more like a state variable. 
```suggestion
           ``process`` part of ``dtend_variable_process``. Processes _Production and Loss Rate_, _Ozone Mixing Ratio_, _Temperature_, and _Overhead Ozone Column_ refer to subprocesses within the ozone parameterization.
```

**Review by ligiabernardet — CHANGES_REQUESTED** [review]
Thank you so much for documenting this capability. It is quite easy to follow.

File: CCPPtechnical/source/ParamSpecificOutput.rst (hunk 4 above)
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst:223]: A comment in the code says, "Ozone production from temperature term at model layers."

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 5]
```diff
@@ -225,7 +343,33 @@ Setting ``ldiag3d=F`` and ``qdiag3d=T`` will result in an error message:
...
+* You did not enable calculation of the individual tendencies, such as ozone. The ``phys_tend`` sums those
+  to make the total tendencies.
+
+Why are my other tendencies zero, even though the model says they're supported for my configuration?
+----------------------------------------------------------------------------------------------------
+
+The tendencies will be zero if they're never calculated. Check that you enabled the tendencies with
+appropriate settings of ``ldiag3d``, ``qdiag3d``, and ``diag_select``. 
+
+Another possibility is that the tendencies in question really are zero. The list of "available" tendencies
```
[the remark this answers is not in the document]
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst:367]: My inability to get a TKE tendency out of satmedmfvdifq is one such gap.

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 6]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
+   | **Process**                    | **diag_table** | **Process**   | **Array Slice**                                            |
+   |                                | **Module**     | **Short**     |                                                            |
+   |                                | **Name**       | **Name**      |                                                            |
+   +================================+================+===============+============================================================+
+   | Planetary Boundary Layer       | ``gfs_phys``   | ``pbl``       | ``dtend(:,:,dtidx(:,index_of_process_pbl))``               |
+   +--------------------------------+----------------+---------------+------------------------------------------------------------+
+   | Deep Convection                | ``gfs_phys``   | ``deepcnv``   | ``dtend(:,:,dtidx(:,index_of_process_dcnv))``              |
+   +--------------------------------+----------------+---------------+------------------------------------------------------------+
+   | Shallow Convection             | ``gfs_phys``   | ``shalcnv``   | ``dtend(:,:,dtidx(:,index_of_process_scnv))``              |
```
[the remark this answers is not in the document]
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst:234]: It only outputs PBL tendency.

File: CCPPtechnical/source/ParamSpecificOutput.rst (hunk 4 above)
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst:223]: I think these are four terms of some equation in the parameterization.

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 7]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
+
+The third step is to enable output of variables from the diag_table, which will be discussed in the next section.
+
+.. _avail_tend_variables:
+
+.. table:: Non-chemical tracer and state variables with tendencies. The second column is the ``variable``
+           part of ``dtend_variable_process``.
+
+   +-------------------------------------------------+----------------+----------------+----------------------------------------------+-------------------------------+
+   | **State Variable Or Tracer**                    | **Variable**   | **Associated** | **Array Slice**                              | **Tendency Units**            |
```
[the remark this answers is not in the document]
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst]: The column heading was unnecessarily long. I have corrected it, and the formatting should be okay now. Please check again.

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 8]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
+   +-------------------------------------------------+----------------+----------------+----------------------------------------------+-------------------------------+
+   | Oxygen Ion Concentration                        | ``o_ion``      | ``qdiag3d``    | ``dtend(:,:,dtidx(100+nto,:))``              | kg kg\ :sup:`-1` s\ :sup:`-1` |
+   +-------------------------------------------------+----------------+----------------+----------------------------------------------+-------------------------------+
+   | Oxygen Concentration                            | ``o2``         | ``qdiag3d``    | ``dtend(:,:,dtidx(100+nto2,:))``             | kg kg\ :sup:`-1` s\ :sup:`-1` |
+   +-------------------------------------------------+----------------+----------------+----------------------------------------------+-------------------------------+
+
+
+.. _avail_tend_processes:
+
+.. table:: Processes that can change non-chemical tracer and state variables. The third column is the
```
[the remark this answers is not in the document]
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst:227]: I added space before each table. How does it look now?

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 9]
```diff
@@ -225,7 +343,33 @@ Setting ``ldiag3d=F`` and ``qdiag3d=T`` will result in an error message:
...
+There are three likely reasons:
+
+* You forgot to enable calculation of physics tendencies. Make sure ``ldiag3d`` and ``qdiag3d`` are T, and
+  make sure ``diag_select`` selects physics tendencies.
+* The suite did not enable the ``phys_tend`` scheme, which calculates the total physics and total
+  photochemistry tendencies.
+* You did not enable calculation of the individual tendencies, such as ozone. The ``phys_tend`` sums those
+  to make the total tendencies.
+
+Why are my other tendencies zero, even though the model says they're supported for my configuration?
```
[the remark this answers is not in the document]
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst]: I swapped the questions and modified the text. Does it look okay now?

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 10]
```diff
@@ -38,139 +38,256 @@ them. It also contains a list of frequently-asked questions in :numref:`Section
...
----------------------
+subdivided by subprocesses: production and loss (combined as a single subprocess), quantity of ozone present
+in the column above a grid cell, influences from temperature, and influences from mixing ratio.  For more
+information about the NRL 2015 ozone photochemistry scheme, consult the `CCPP Scientific Documentation
+<https://dtcenter.ucar.edu/GMTB/v5.0.0/sci_doc/GFS_OZPHYS.html>`_.
+
+There are three steps of selecting the tendencies to output: enable diagnostics, select which tendencies to
+calculate, and select which ones to output. To determine what tendencies are available for your
+configuration, enable tendencies, but select only one of them, as discussed later. (Non-physics temperature
+tendency is available for all suites.) Then rerun with the desired tendencies enabled.
```
[the remark this answers is not in the document]
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst]: I rewrote that, and rearranged the later code. It is now explained better.

File: CCPPtechnical/source/ParamSpecificOutput.rst (hunk 4 above)
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst]: I updated the tables to clarify that.

**Comment by SamuelTrahanNOAA** [comment]
> Thank you so much for documenting this capability. It is quite easy to follow.

Hopefully it is still easy to follow now that I've rearranged it.

File: CCPPtechnical/source/ParamSpecificOutput.rst [hunk 11]
```diff
@@ -179,15 +296,15 @@ by variable ``max_output_fields`` in namelist section ``&diag_manager_nml`` in f
...
 
 SCM
 ^^^
 
-The default behavior of the SCM is to output instantaneous values of all variables in
-:numref:`Table %s <avail_tendencies>`. Tendencies are computed in file ``gmtb_scm_output.F90`` in
-the subroutines output_init and output_append. If the values of ``ldiag3d`` or ``qdiag3d`` are set
-to false, the variables are still written to output but are given missing values.
+The default behavior of the SCM is to output instantaneous values of all tendency variables, and
+``dtend_select`` is not recognized. Tendencies are computed in file ``gmtb_scm_output.F90`` in the
```
[the remark this answers is not in the document]
Comment (SamuelTrahanNOAA) [reply CCPPtechnical/source/ParamSpecificOutput.rst:331]: The logic to handle `dtend_select` globbing (`*` and `?` characters) is not model-specific. That's the `fglob_list` and `fglob` functions. Grant can lift it out of GFS_typedefs.F90 and put it in SCM. The fill_dtidx and labelling logic use the model data structures, but that shouldn't be hard to fix.

I've never looked at the SCM code though. If you wrote it in Ada or C++, you've a bit more work ahead of you.

File: CCPPtechnical/source/ParamSpecificOutput.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (JulieSchramm) [reply CCPPtechnical/source/ParamSpecificOutput.rst]: The table header looks good to me.

File: CCPPtechnical/source/ParamSpecificOutput.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (ligiabernardet) [reply CCPPtechnical/source/ParamSpecificOutput.rst]: Yes, looks good now.

File: CCPPtechnical/source/ParamSpecificOutput.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (ligiabernardet) [reply CCPPtechnical/source/ParamSpecificOutput.rst:331]: The SCM is in Fortran so your development in GFS_typedefs.F90 should be easy to transfer to the SCM.

File: CCPPtechnical/source/ParamSpecificOutput.rst (hunk 4 above)
Comment (ligiabernardet) [reply CCPPtechnical/source/ParamSpecificOutput.rst]: Looks good

**Review by ligiabernardet — APPROVED** [review]
Documentation looks good.
One @grantfirl has this capability working in the SCM, some touch ups will be needed. Including being clear about SCM using time-aggregated tendencies (doc still says instantaneous).

**Comment by SamuelTrahanNOAA** [comment]
Can this PR be merged now? The code PRs are all merged to their respective central repositories.

**Pull request merged**